### ISSUE TITLE
Implementing Marthe for other models - location of last code line in training loop?

### ISSUE THREAD

**Issue by lessw2020** [awslabs/adatune#4]
Hi - quite excited about your work here on hypergradient optimization.

I ran the baselines with VGG to test things out and now trying to setup MARTHE for my own training loop with an effdet object detector.  I'm not able to get any actual learning to take place so I want to clarify the placement esp of the last line of your implemention block.

The readme states to implement:

> first_grad = ag.grad(loss, net.parameters(), create_graph=True, retain_graph=True)
> hyper_optim.compute_hg(net, first_grad)
> for params, gradients in zip(net.parameters(), first_grad):
>      params.grad = gradients
> optimizer.step()

> hyper_optim.hyper_step(vg.val_grad(net))  #<-- last line

my question is where  does the last line go - literally right after optimizer.step() as shown above?  or is it moved below to after the validation step is done?
>        if epoch % test_interval == 0:
>            model.eval()
If I move it to where I think it should go (after eval) then it complains about vg not defined...which is another question, where is vg object coming from?)

Possibly related to my issue - I get this error on every mini-batch as well:
One of the differentiated Tensors appears to not have been used in the graph. Set allow_unused=True if this is the desired behavior.

I'll re-read the paper and see what I might have missed but confirming the code placement above would be appreciated (as well any other guidelines so I can get this up and running).
Thanks!

**Comment by lessw2020** [comment]
ok I see in rtho.py that you are creating vg with:
>vg = ValidationGradient(test_data, nn.CrossEntropyLoss(), device)

and then going into your training loop...so the readme code block is correct for those lines, but isn't quite reflective of the fact you need to make an initial vg prior to going into your training loop.

I should be able to put it together now that I've found this but might be helpful to clarify implementation in a bit more detail.  
The readme should add the above line, and perhaps the additional note of:
>hyper_optim = MuSGD(optimizer, hyper_lr, grad_clipping, first_order, mu, alpha, device)

Thanks!
